tests/jni_api.rs

```diff
@@ -315,122 +315,140 @@ pub fn java_byte_array_from_slice() {
-
-        // Check pointer into() access
-        let ptr: *mut jbyte = (&auto_ptr).into();
-        assert_eq!(unsafe { *ptr.offset(0) }, 1);
-        assert_eq!(unsafe { *ptr.offset(1) }, 2);
-        assert_eq!(unsafe { *ptr.offset(2) }, 3);
-
-        // Modify
-        unsafe {
-            *ptr.offset(0) += 1;
-            *ptr.offset(1) += 1;
-            *ptr.offset(2) += 1;
+macro_rules! test_get_array_elements {
+    ( $jni_get:tt, $jni_type:ty, $new_array:tt, $get_array:tt, $set_array:tt ) => {
+        #[test]
+        pub fn $jni_get() {
+            let env = attach_current_thread();
+
+            // Create original Java array
+            let buf: &[$jni_type] = &[1 as $jni_type, 2 as $jni_type, 3 as $jni_type];
+            let java_array = env
+                .$new_array(3)
+                .expect(stringify!(JNIEnv#$new_array must create a Java $jni_type array with given size));
+
+            // Insert array elements
+            let _ = env.$set_array(java_array, 0, buf);
+
+            // Use a scope to test Drop
+            {
+                // Get byte array elements auto wrapper
+                let auto_ptr: AutoArray<$jni_type> =
+                    env.$jni_get(java_array, ReleaseMode::CopyBack).unwrap();
+
+                // Check array size
+                assert_eq!(auto_ptr.size().unwrap(), 3);
+
+                // Check pointer access
+                let ptr = auto_ptr.as_ptr();
+                assert_eq!(unsafe { *ptr.offset(0) } as i32, 1);
+                assert_eq!(unsafe { *ptr.offset(1) } as i32, 2);
+                assert_eq!(unsafe { *ptr.offset(2) } as i32, 3);
+
+                // Check pointer From access
+                let ptr: *mut $jni_type = std::convert::From::from(&auto_ptr);
+                assert_eq!(unsafe { *ptr.offset(0) } as i32, 1);
+                assert_eq!(unsafe { *ptr.offset(1) } as i32, 2);
+                assert_eq!(unsafe { *ptr.offset(2) } as i32, 3);
+
+                // Check pointer into() access
+                let ptr: *mut $jni_type = (&auto_ptr).into();
+                assert_eq!(unsafe { *ptr.offset(0) } as i32, 1);
+                assert_eq!(unsafe { *ptr.offset(1) } as i32, 2);
+                assert_eq!(unsafe { *ptr.offset(2) } as i32, 3);
+
+                // Modify
+                unsafe {
+                    *ptr.offset(0) += 1 as $jni_type;
+                    *ptr.offset(1) += 1 as $jni_type;
+                    *ptr.offset(2) += 1 as $jni_type;
+                }
+
+                // Commit would be necessary here, if there were no closure
+                //auto_ptr.commit().unwrap();
+            }
+
+            // Confirm modification of original Java array
+            let mut res: [$jni_type; 3] = [0 as $jni_type; 3];
+            env.$get_array(java_array, 0, &mut res).unwrap();
+            assert_eq!(res[0] as i32, 2);
+            assert_eq!(res[1] as i32, 3);
+            assert_eq!(res[2] as i32, 4);
```

Comment:
Isn't that problematic for floats?

Reply:
That's there because float equalities are unreliable, and require a precision parameter. It's that, or introducing a precision parameter for those equalities.

This is just a way to "overload" the test case to work for floats. I can also separate float tests cases for clarity and  "correctness", but I don't think it's really necessary. As you wish.

Reply:
OK, let's keep it as is, if these casts and comparisons work.